### PR TITLE
Refactor Geohash Grid Aggregation

### PR DESCRIPTION
**Pull request by colings86**

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoHashGridParser.java [hunk 1]
```diff
@@ -141,6 +155,23 @@ public InternalAggregation buildEmptyAggregation() {
         protected Aggregator doCreateInternal(final ValuesSource.GeoPoint valuesSource, AggregationContext aggregationContext,
                 Aggregator parent, boolean collectsFromSingleBucket, List<PipelineAggregator> pipelineAggregators, Map<String, Object> metaData)
                 throws IOException {
+            if (shardSize == 0) {
+                shardSize = Integer.MAX_VALUE;
+            }
+
+            if (requiredSize == 0) {
+                requiredSize = Integer.MAX_VALUE;
```
Comment (nknize) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoHashGridParser.java]: I was curious why this is set to ```Integer.MAX_VALUE```? Would we not want to reset it to ```DEFAULT_MAX_NUM_CELLS```?

File: core/src/test/java/org/elasticsearch/search/aggregations/bucket/GeoHashGridIT.java [hunk 2]
```diff
@@ -136,6 +135,7 @@ public void setupSuiteScopeCluster() throws Exception {
     @Test
     public void simple() throws Exception {
         for (int precision = 1; precision <= XGeoHashUtils.PRECISION; precision++) {
+            System.out.println("precision: " + precision);
```
Comment (nknize) [inline core/src/test/java/org/elasticsearch/search/aggregations/bucket/GeoHashGridIT.java]: Sneaky System.out

Comment (colings86) [inline core/src/test/java/org/elasticsearch/search/aggregations/bucket/GeoHashGridIT.java]: Good catch. I'll remove

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoHashGridParser.java (hunk 1 above)
Comment (colings86) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoHashGridParser.java]: So we have a loose convention that `size: 0` means get all buckets. I personally don't like it but we have to keep it for now I guess as this refactor should be backwards compatible

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoHashGridParser.java [hunk 3]
```diff
@@ -141,6 +155,23 @@ public InternalAggregation buildEmptyAggregation() {
...
+
+            if (shardSize < 0) {
+                // Use default heuristic to avoid any wrong-ranking caused by
+                // distributed counting
+                shardSize = BucketUtils.suggestShardSideQueueSize(requiredSize, aggregationContext.searchContext().numberOfShards());
+            }
+
+            if (shardSize < requiredSize) {
+                shardSize = requiredSize;
+            }
```
Comment (nknize) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoHashGridParser.java]: so if ```shardSize == -1```, and ```requiredSize == 0```, then ```shardSize``` will be reset here to ```Integer.MAX_VALUE```?

Comment (colings86) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoHashGridParser.java]: yes

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoHashGridParser.java [hunk 4]
```diff
@@ -21,108 +21,118 @@
...
+        GeoGridFactory factory = new GeoGridFactory(aggregationName);
+        Integer precision = (Integer) otherOptions.get(GeoHashGridParams.FIELD_PRECISION);
+        if (precision != null) {
+            factory.precision(precision);
                 }
-            } else if (token != XContentParser.Token.START_OBJECT) {
-                throw new SearchParseException(context, "Unexpected token " + token + " in [" + aggregationName + "].",
-                        parser.getTokenLocation());
+        Integer size = (Integer) otherOptions.get(GeoHashGridParams.FIELD_SIZE);
+        if (size != null) {
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoHashGridParser.java]: indentation looks wrong?

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoHashGridParser.java [hunk 5]
```diff
@@ -21,108 +21,118 @@
...
+        }
 
-    static class GeoGridFactory extends ValuesSourceAggregatorFactory<ValuesSource.GeoPoint> {
-
-        private final int precision;
-        private final int requiredSize;
-        private final int shardSize;
+        public void size(int size) {
+            this.requiredSize = size;
+    }
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoHashGridParser.java]: indentation

**Comment by jpountz** [comment]
LGTM, just found some indentation issues

**Pull request merged**